### PR TITLE
Use pre-commit

### PR DESCRIPTION
**Pull request by saimn**
I was looking for a way to run checks for the changelog fragment filenames and pre-commit seems a good option. Actually tox already has such a check so I adapted theirs. With this pre-commit will check that the fragments have one of the correct categories (bugfix|feature|api|other), and that files with the "other" category are put only in the root directory (following https://github.com/astropy/astropy/pull/11856#discussion_r654565093).

I also added some basic hooks for trailing whitespace and end of lines (@Cadair proposed a similar config in #9890)
About flake8: I added a hook to run flake8 via `tox -e codestyle`. In #9890 @Cadair did it another way by running pre-commit from tox. After playing with this, I think we should remove flake8 / codestyle from the tox config and do it directly with pre-commit, there is no real benefit in having the possibility to run flake8 with both. And since pre-commit is about running code style checks, flake8 should be part it.

If we go with this, we will need to run it once to fix the existing files, I can do that in this PR later (after 4.3 is out and fragments have been deleted in main).

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  - id: trailing-whitespace
+    exclude: ".*(data.*|extern.*|licenses.*|_parsetab.py)$"
+# - repo: https://github.com/timothycrosley/isort
+#   rev: 5.9.2
+#   hooks:
+#   - id: isort
+# - repo: https://github.com/asottile/pyupgrade
+#   rev: v2.21.0
+#   hooks:
+#     - id: pyupgrade
```
Comment (saimn) [inline .pre-commit-config.yaml:20]: 2 additional checks we could use: 
- we have an isort config, should we run it (I thought we ran it once in the past but actually it seems not)
- pyupgrade was used to convert to fstrings in #8958, it can also provide various little improvements (https://github.com/asottile/pyupgrade) using syntax from the recent versions of Python.

**Comment by saimn** [comment]
@pllim - No, it just runs on all changelog fragment, and cannot check that a PR actually has a fragment. And it couldn't be skipped with some label.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (nstarman) [reply .pre-commit-config.yaml:20]: Other good built-in ones (v3.4.0>) are:

check-added-large-files
check-case-conflict
check-merge-conflict
check-symlinks
check-toml
check-yaml
debug-statements
destroyed-symlinks
detect-private-key
end-of-file-fixer
fix-encoding-pragma
requirements-txt-fixer
trailing-whitespace

Comment (nstarman) [reply .pre-commit-config.yaml:20]: > isort config, should we run it

👍

Comment (nstarman) [reply .pre-commit-config.yaml:20]: And https://github.com/pre-commit/pygrep-hooks is great for rst checks.

- rst-backticks
- rst-directive-colons
- rst-inline-touching-normal

I also use it for checking that I give the right error message in ``noqa``  e.g. ``noqa: F403``

- python-check-blanket-noqa

**Comment by bsipocz** [comment]
minor idea: maybe this should also check whether the changelog is too wordy, e.g. notify the reviewer to double-check the entry if it's longer than ~5 lines?

**Review by pllim — COMMENTED** [review]
Instead of adding all sorts of checks in this PR, perhaps we should go with what is already here and defer the rest for follow-up PRs.

Release managers should carefully review the change log check, as it will affect them the most at release time.

**Comment by saimn** [comment]
> minor idea: maybe this should also check whether the changelog is too wordy, e.g. notify the reviewer to double-check the entry if it's longer than ~5 lines?

I guess I could make it fail if a fragment has more than 5 lines, but then with the way it works it would always fail until you fix that file. So it would prevent having exceptions unless adding the file to the exceptions in pre-commit's config, which would not be practical.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (saimn) [reply .pre-commit-config.yaml:20]: > Other good built-in ones (v3.4.0>) are:

I already added most of them:
```
  - id: check-added-large-files
  - id: check-case-conflict
  - id: check-yaml
  - id: debug-statements
  - id: end-of-file-fixer
  - id: trailing-whitespace
```

**Comment by bsipocz** [comment]
>  it works it would always fail until you fix that file. 

what about a new special label? :)

**Comment by saimn** [comment]
pre-commit does not access github labels and does not know anything about the current PR. It just runs a number of checks on a selection of files. So for the changelog fragments I just tell it to check that files in `doc/changes/` respect some rules, with a regex.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (nstarman) [reply .pre-commit-config.yaml:20]: Looks like they are working! The CI has found oodles of corrections in the code.

Comment (pllim) [reply .pre-commit-config.yaml:20]: Is "oodles" good or bad? 😅

Comment (nstarman) [reply .pre-commit-config.yaml:20]: Both! 😆

**Comment by nstarman** [comment]
I checked out this branch and ran ``pre-commit run --all-files`` twice. I can confirm that the second run produced no errors, except for the expected error of misplaced / misnamed "other" changelog fragments.

**Comment by nstarman** [comment]
Rerunning the CI to see if the code style check job error is real...
